### ISSUE TITLE
tpm2_create issue for RSA/ECC keys with restricted object attribute

### ISSUE THREAD

**Issue by mrcdb** [tpm2-software/tpm2-tools#1229]
I am trying to create some ECC keys in the owner hierarchy with the 3.1.3 release of the tool.

First, I create a restricted encryption key as follows and store its context:
```
$ tpm2_createprimary -H o -g sha256 -G ecc -C primary/context.out
```
The command succeeds, and I get an handle as well.
Then, I try to  generate either child restricted asymmetric signing and decryption keys with the following commands:
```
$ tpm2_create -c primary/context.out -g sha256 \
    -G ecc -u sign/child.pub -r sign/child.priv \
    -A "sign|fixedtpm|fixedparent|restricted|sensitivedataorigin|userwithauth"
$ tpm2_create -V -c primary/context.out -g sha256 \
    -G ecc -u dec/child.pub -r dec/child.priv \
    -A "decrypt|fixedtpm|fixedparent|restricted|sensitivedataorigin|userwithauth"
```
In both cases, I get error code `0x2c2`, which means `inconsistent attributes`. What am I doing wrong?

If I use the same object attributes and use a `-G keyedhash` algorithm option, key generation succeeds. If I generate child **non-restricted** signing/encryption keys, it works as well.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Marco De Benedictis [mailto:notifications@github.com]
> Sent: Monday, November 5, 2018 9:49 AM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Subscribed <subscribed@noreply.github.com>
> Subject: [tpm2-software/tpm2-tools] tpm2_create issue for RSA/ECC keys with
> restricted object attribute (#1229)
> 
> I am trying to create some ECC keys in the owner hierarchy with the 3.1.3 release
> of the tool.
> 
> First, I create a restricted encryption key as follows and store its context:
> 
> $ tpm2_createprimary -H o -g sha256 -G ecc -C primary/context.out
> 
> 
> The command succeeds, and I get an handle as well.
> Then, I try to generate either child restricted asymmetric signing and decryption
> keys with the following commands:
> 
> $ tpm2_create -c primary/context.out -g sha256 \
>     -G ecc -u sign/child.pub -r sign/child.priv \
>     -A "sign|fixedtpm|fixedparent|restricted|sensitivedataorigin|userwithauth"
> $ tpm2_create -V -c primary/context.out -g sha256 \
>     -G ecc -u dec/child.pub -r dec/child.priv \
>     -A
> "decrypt|fixedtpm|fixedparent|restricted|sensitivedataorigin|userwithauth"
> 
> 
> In both cases, I get error code 0x2c2, which means inconsistent attributes. What
> am I doing wrong?
> 
> If I use the same object attributes and use a -G keyedhash algorithm option, key
> generation succeeds. If I generate child non-restricted signing/encryption keys, it
> works as well.

A bit of lore for 3.X:
https://github.com/tpm2-software/tpm2-tools/commit/352908f7470338aba858a9ac20c1c5ec608b661c

So you're attributes are getting mixed in with stuff:
#define DEFAULT_ATTRS \
     TPMA_OBJECT_DECRYPT|TPMA_OBJECT_SIGN_ENCRYPT|TPMA_OBJECT_FIXEDTPM \
    |TPMA_OBJECT_FIXEDPARENT|TPMA_OBJECT_SENSITIVEDATAORIGIN \
    |TPMA_OBJECT_USERWITHAUTH

A check in the TPM simulator in function PublicAttributesValidation() checks the following:
 
    // See if sign and decrypt are the same
    if(attributes.sign == attributes.decrypt)
	{
	    // a restricted key cannot have both SET or both CLEAR
	    if(attributes.restricted)
		return TPM_RC_ATTRIBUTES;

and your falling into this with a restricted key. However, even if you overcome this, you'll get to the scheme checks and die, see routine SchemeChecks()
in the simulator.

I jumped down the rabbit hole for an hour or so looking at a way we could do it, it involved:
1. Using a dash in the front of attributes to infer the memset 0 and use the user supplied attrs as is
2. Introspecting attrs and inferring how to set up the symmetric details for ECC and Signing scheme based on decrypt/encrypt

However, then I hit a check ensuring that the child object and non-duplicable parent objects symmetric details match:
	    
                     // If this parent is not duplicable, then the symmetric algorithms
	    // (encryption and hash) must match those of its parent
	    if(attributes.fixedParent && (parentObject != NULL))
		{
		    if(publicArea->nameAlg != parentObject->publicArea.nameAlg)
			return TPM_RCS_HASH;
		    if(!MemoryEqual(symAlgs, &parentObject->publicArea.parameters,
				    sizeof(TPMT_SYM_DEF_OBJECT)))
			return TPM_RCS_SYMMETRIC;

I then tried turning off fixedparent|fixedtpm, and fell into other checks..... This is one my least favorite parts of the TPM. They made this stuff
ETOOCOMPLICATED IHMO.

The patch attached does get this to work:
tpm2_create -Tmssim -c primary.ctx -g sha256 -G ecc -u child.pub -r child.priv -A "-decrypt|restricted|userwithauth|sensitivedataorigin"

Note: I had to remove fixedparent and fixedtpm which is NOT ideal, those object attributes might not work for you.

In short this stuff is way better supported on master, I'd recommend using that branch. You can specify all the bits of the object creation
as needed via -A and -G.

On master your commands are:
tpm2_createprimary -Gecc -o primary.ctx
tpm2_create  -C primary.ctx -g sha256 -G ecc -u child.pub -r child.priv -A "decrypt|fixedparent|fixedtpm|restricted|userwithauth|sensitivedataorigin"

I'd take patches that correct this in a backwards compatible way on 3.X. I attached my work as attrs.patch, feel free to use that as
desired.

Bill

**Comment by williamcroberts** [comment]
Please use the mailing list for support.
